Update main branch with current code

Bryan0x05 opened the pull request:
Test branch is a super set of all branches, it contains:
- Unit test layer
 - basic network test
 - stubs for future encryption support
- Functional basic networking
- Stub/Skeleton Encryption lib
- Refactor code into 3 core libs
- Tmux server manager


I would also like to merge this then prune all the other branches to keep things manage-able. Also sorry for the large merge, main branch is really behind.

Bryan0x05 committed:
push

Bryan0x05 committed:
Checkpoint upload

Add a stub messageHandler class, corrected some socket logic in netlib, added some cmdloop logic to client.

Fixed processes to run parallel

issues:
- cmd.loop dislikes being run from inside a new process
- socket connections seem to drop

Bryan0x05 committed:
MessageHandler update

Finish messageHandler,
Update client-server logic to use messageHandler,
test the code

Bryan0x05 committed:
Updated MsgHandler + Threading

We added the argument separators to the message handler.

Then, we adjusted the readMsg, sendMsg and several commands in the shell class to use messageHandler.

There were several updates to threadplus class. Right now multi-threading does not play nice with the cmdloop + listen loop.

Bryan0x05 committed:
Cmdline + Server Logic Update

For now the client listener thread is commented out.

Updated cmdline logic, to use send messageHandler processed messages to the server and some limited server logic to process and act on those messages.

Fixed some bugs like in readMsg as well.

todo:
Figure out what to do with the listener thread, so we can replies.

tom-kap committed:
Update netlib.py

Bryan0x05 committed:
sync commit

commit for team members to recreate problems

Bryan0x05 committed:
Bug fix + limited client logic

This + the last couple of commits have fixed:

- Cmd and listen loop multithreads no longer fight ( We aren't 100% how )
- Added limited server & client back-and-forth
-- Client now updates its nickname dictionary when listSockets is called
- Other misc bug fixes and code cleanups

TODO:
- Fully implement client <-> server networking cases and logic.
- Fully implement node <-> node logic
- write basic networking tests uses process pairs ( NOTE: cmdloop(), might throw a fit in another process, that hasn't been addressed yet )

Bryan0x05 committed:
-Fleshed out stubs
-Added more server <> client case
-Added some cases for node <> node ( Not tested for now )
-Code tidying
-- Changed some code segments to use existing functions with error handling
-- changed read/send message functions to static methods

Bryan0x05 committed:
-Combined server and client into peer
-Finished local logic
-Started work on node <> node communcation, currently stdin/stdout
from one terminal attempts to control everything. Need a work around.

Bryan0x05 committed:
A lot of changes, bug fixes and new bugs.
A lot of peer functionality has been moved to a new start function. This
start function can start a sub-process, and redirects input and output.

Peer now has wrappers, so a top level function, can use these wrappers
to pipe commands to the shell class and interactive with the peer object
inside the sub process. These functions work as in shell registers and
runs these commands, but its hard to validate behavor without direct access
to the object.

Shifted around some files, to now have a manSetup.py file to show whole
the new start works, as well as unitTest file that also leverages the
new subprocess functionality for testing.

Some known bugs(likely mainly more):
- Send Msg command runs, but no msg is seen on the output of the other
  peer.
- In manSetup.py, connecting to peer2, then trying to quit results in
  quit hanging. Unknown if this is a bug with connection or what.

Bryan0x05 committed:
- Strengthen test case for basic peer <> peer networking
- Added closeConn function

Bryan0x05 committed:
Added test suite to github actions(hopefully this is a test push to see
if it does work)

Bryan0x05 committed:
fixed minor bug of using wrong yaml key word

Bryan0x05 committed:
Expanding github action triggers to include all branchs

Bryan0x05 committed:
What's new:
-Python script to create a split terminal window, each with their own
peer
-- this uses tmux which is present on the cs servers (though the module
to manage this might need to be installed )
- Some additional logic to support needed run state for the tmux stuff
- bug fixes

Bryan0x05 committed:
- minor tweaks and code clean-up.
This was meant to have sendMsg / recvMsg testing. But sockets are
behaving oddly in the test suite, comments detail the specific odd
behaviors.

Testing might be better if switched over to using the tmuxcode but I
think the current level of testing is fine.

Bryan0x05 committed:
Some minor code clean-up

Bryan0x05 committed:
* CS Server has Python 3.8.10, tmuxSetup.py uses a python 3.9+ version of libtmux
Decided it was easier to make a bash script instead of backporting ( see tmuxSetup.sh not .py)
* Some terminal formatting
* Removed some user-visible(non-logged) debug statements

Bryan0x05 committed:
* Co-authored-by: Tom Kaplan <tom-kap@users.noreply.github.com>
* Fixed readme formatting

Bryan0x05 committed:
Summary

Currently this code is broken, because I'm in the middle of altering the
acceptConn and connectToIp logic to make a true 2-way connection right
now we have a bit of logical bug making it one-way because we are
peer2peer. I decided it might be best to finish fixing with a second
pair of eyes to not spawn a bunch future bugs off of th fix.

Summary of my other changes over several commits:
- Color coded messages
 - Red = error, green = command print msg, blue = message from peer
- Test Suite orgaization
- Alternative bash script to manage tmux since libtmux module is dated
  on mercury
- Clear screen command on terminal
- Lan ip address function ( no more using ip 0.0.0.0 or 127.0.0.1 )
 - This should mean someone can connect to us from outside our machine
   in theory now.
 - Updated readme
  - its now semi-comphrensive
 - setupNode.py
  - additional cmdline argument value to toggle logger prints ( false by
    default)

Suggestions:
- setup Python 3.8.1 virtual environment to match the cs server python
  version
-- Match needs to be converted to a series of else ifs
-- tmuxSetup.py will be defunct ( replaced by tmuxSetup.sh )
-- Additional flag / cmdline argument in setupNode.py to

jojojo8359 committed:
Fix 3.8.10 issues

- Fix typing for this version - use Tuple and List instead of tuple and list
- Convert match statement to if-else block

Bryan0x05 committed:
Attempted 3-way handshake( still broken T.T)

Bryan0x05 committed:
Merge pull request #9 from Bryan0x05/p2pd

Attempted 3-way handshake( still broken T.T)

Bryan0x05 committed:
Summary

- fixed 3-way handshake
- re-factored code into 3 core lib files
- Added security manager with PKC generation
- Added a stub uris dictionary ( for uni resource identify for peers)
  - These are meant to be unique network wide.

Bryan0x05 committed:
test hotfix
- test now works and tests the new 2-way communcation between 2 peers

Bryan0x05 committed:
whoops, left an unused import in requirements.txt ( was used but then
given up on)

Bryan0x05 committed:
some more fixing based on pyright feedback from gitactions

Bryan0x05 committed:
Hopeful solution a niche pywright case git actions is getting suck on
that I do not get locally with pylance

Bryan0x05 committed:
Another attempted pyright fix, it really shouldn't be a problem, it
can't ever be none but it thinks it can be none

Bryan0x05 committed:
Solution #3

Bryan0x05 committed:
more type: ignores

Bryan0x05 committed:
Mended github action for test file to include the new test file name
that works better with the intergrated VSC tester

Bryan0x05 committed:
fixed typo

Bryan0x05 committed:
More skeleton functions in seclib to support public key infrastructure we'll need later.
0% tested, and is currently still un-used

Bryan0x05 committed:
Fixed seclib to actually use RSA public/private key pairs instead of a
symmetric encryption like AES which is not pub/pri keys(my bad).

Bryan0x05 committed:
Updating comments to reflect RSA is now the encryption algo

Bryan0x05 committed:
Summary:
-RSA to replace AES, since AES is not PKC.
-Merged RSA changes to testbranch version that on was on git with key ring.
-Updated Comments and imports to reflect this

Bryan0x05 committed:
-Replaced AES with RSA, since AES is not PKC
-Resolved merge conflict of RSA change on testBranch

Bryan0x05 committed:
Ignored pywright false-positive with another cryptography function call

Bryan0x05 committed:
Elevated Pyright ignore to generatePKCKeys function level.

It really should not be picking up the optional backend parameter that is not being invoked.

Bryan0x05 committed:
More pyright ignores

Bryan0x05 committed:
- Set python 3.8.10 for github actions given that's our target version
- Another attempt a pyright fix, odd it worked with previous backend
  issue but not this one.

jojojo8359 commented on .github/workflows/main.yml:
```diff
@@ -1,28 +1,36 @@
+
+jobs:
+  test:
+    runs-on: ubuntu-latest
```
Looks like we might need to change this to "ubuntu-20.04" for Python 3.8.10 to work.

jojojo8359 commented on README.md:
```diff
@@ -1 +1,20 @@
+
+## requirements
+* Linux ( Ubuntu )
+* Python >= 3.8.1
```
Change to 3.8.10?

jojojo8359 reviewed (commented):
Just a couple of things:
- Potential fix for actions not running
- Quick update to README

Bryan0x05 committed:
- CONNECTTOTIP NOW USES SPIN AMINATION!
- Github actions yaml updated w/ Joel's suggestions
- Added peer shutdown command + updated/debugged shutdown logic
- Cleaned-up shell output when using connectToIp
- Added a timeout to connectToIp ( default timeout is very long )

Bryan0x05 committed:
Adding the yaml changes that should've been in the prior commit

Bryan0x05 committed:
Merge branch 'testBranch' of https://github.com/Bryan0x05/c5610Project into testBranch
Merging, 2 different versions of testBranch together.

Summary:
- Cleaner output
- Faster timeout on connectToIp
- Github Action YAML updated
- Node shutdown logic + shutdown command

Bryan0x05 committed:
- CA added
- CA commands
- Peer Commands

todo:
- Peer needs to send and read CA messages
- Read and send Msg need to encrypt/decrypt conditionally

Bryan0x05 committed:
Merge branch 'testBranch' of github.com:Bryan0x05/c5610Project into testBranch

see last summary, merge conflicts were resolved this commit

Bryan0x05 committed:
- Added conditional en/de-crypt in send and read messages.
- Squashed bugs
- Fixed unit test to use 3-way handshake
- Fixed yaml typo in ubuntu runner image

Bryan0x05 committed:
More stubs for CA logic + code cleanup

Bryan0x05 committed:
- Added pyenv to github workflow
-- Hopefully this allows us to use python3.8.10
- Added conditional encrypt/decrypt logic to send and read msg
- Refactored xchange_key code, and placed it into a function
- Added encrypted test case and refactored existing test code for code-reusability between tests
-- NOTE: tests are currently failing, refactoring likely introduced a bug or logic fault.

jojojo8359 committed:
Fix broken actions

- No default was provided in action, and we don't need any additional setup within the pyenv action

Bryan0x05 committed:
- Updated / Fix unittesting. Should have converage over all currently
  implemented logic ( i.e. anything but the CA)
- Renamed runLoop to runCycle, since it doesn't internally loop
- Renamed SHUT_DWN command to SHUT_DWNCON to make it clear it only
  affects a connection
- Moved SHUT_DWNCON logic to function

Bryan0x05 committed:
- Fleshed CA logic out ( still a WIP )

Bryan0x05 committed:
Added warning if a msg is from an uncerted peer, if we have own our cert(overload as a flag)

Bryan0x05 committed:
we broke encryption, send help

Bryan0x05 committed:
- Fixed exchange key
-- Duplicate functions in clilib cleaned up
-- Faulty if statement printed failure message even on success (
explains why it passed automated tests before but not manual )
-- Adjusted base64 / encode order
- Added "low-level readMsg"
-- Used by acceptConn, it makes far less assumptions and doesn't try
dictionaries lookup
- Cleaned up handshake logic
-- Flow should be a bit clearer
-- No longer making redudant dictionary entries
- Adjusted CA logic
-- Added some compression logic to CA to enable it to encrypt same
length RSA key
-- TODO: Handle the implications of encrypting
-- Added a comparsion function to keyring for the CA

Bryan0x05 committed:
Added uri logic and auto connected logic but broke nicknames dictionary updates

Bryan0x05 committed:
cleaned-up code for demo-ing, fixed request/send URI, TODO: fix quit with 2+ connections

Bryan0x05 committed:
exchange key is broke, CA reg key also broke, check_key broke

Bryan0x05 committed:
fixed exchange key, by introducing new low level protoheader format

Bryan0x05 committed:
Expanded CA logic, need to validate checkkey logic

Bryan0x05 committed:
various fixes, isloated issue with 2+ connections, socket module accept call re-uses the same socket for inSock which causes havoc for us

Bryan0x05 committed:
Fixed inbound socket, sockets aren't insured to have an unique ipAddr:Port, several sockets can have the same one in fact. This hotfix fixes this false assumption

Bryan0x05 committed:
on-going code clean-up and adjustments. Still having issues with CA certifying peers.

Bryan0x05 committed:
Checkpoint push, final fixes are in progress
- Partial code cleanup, logic and style
- CA no longer prints PEER CLI prints ( or any non log print for that
  matter)
- Near final implementation auth code
- Switch to sha224, meaning hashses produce are smaller, and RSA key
  size restriction is a few bytes larger. This is to bypass a size issue
  of when the peer tries to send a CERT+URI to the CA as its just enough
  bytes combined to go over when using sha256.
- Tracking down the final issues with CA cert validation requests not
  being received by the CA

Bryan0x05 committed:
checkpoint upload, working code, updated readme & docstr, tidying still WIP

Bryan0x05 committed:
more bug fixes, loose ends tied up, format fixes, document improvements
